### PR TITLE
WIP: Fix #1385: Upgrade to webpack 5

### LINKED ISSUE #1385: Migrate to webpack version 5
Since webpack 5 is stable now (no longer RC), we should migrate at some point soon-ish.

We need to adjust the related build scripts:
https://github.com/neomjs/neo/tree/dev/buildScripts/webpack

Plus double-check, if the external plugins we are using still work.

I am flagging the ticket as a "good first issue", since you can work on it without having knowledge about this framework.

I am also flagging it as "help wanted", since I am busy with polishing the new drag&drop implementation.

### PR DESCRIPTION
**Pull request by h1b9b**
Simple PR that upgrades webpack to version 5.
There is only one change is in the `server-start` script. 
Instead of using `webpack-dev-server --open`, we use the webpack cli directly `webpack serve --open`.

All tests pass and I didn't see any regression while browsing the example apps

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [x] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch, _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)

### CODE REVIEW

File: buildScripts/docs/jsdocx.js [hunk 1]
```diff
@@ -376,12 +376,15 @@ jsdocx.parse(options)
...
             path = key.replace(/\./g, '/');
-
-            helper.writeJSON({
-                path  : './docs/output/src/' + path + '.json',
-                indent: 4,
-                force : true
-            }, ns('Neo.' + key));
+            const struct = ns('Neo.' + key);
+
+            if (struct) {
```
Comment (h1b9b) [inline buildScripts/docs/jsdocx.js:381]: This is a quick fix to a warning popping when building the docs:
```
(node:21554) UnhandledPromiseRejectionWarning: TypeError [ERR_INVALID_ARG_TYPE]: The "data" argument must be of type string or an instance of Buffer, TypedArray, or DataView. Received undefined
    at writeFile (fs.js:1436:5)
    at go$writeFile (./node_modules/graceful-fs/graceful-fs.js:139:14)
    at Object.writeFile (./node_modules/graceful-fs/graceful-fs.js:136:12)
    at ./node_modules/jsdoc-x/node_modules/universalify/index.js:13:12
    at new Promise (<anonymous>)
    at Object.writeFile (./node_modules/jsdoc-x/node_modules/universalify/index.js:7:14)
    at ./node_modules/jsdoc-x/src/lib/helper.js:119:28
```

**Comment by tobiu** [comment]
Hi @h1b9b! Thx for the PR, appreciated.

I do have time tomorrow to take a deeper look into it.

Did you check if the WebpackShellPlugin still works? (buildThreads => App)
The last time I tested it, it broke (quite a while ago though).

It would not be a big deal in case the plugin does not support webpack v5. We don't really need the onBuildExit hook => we could move the docs resources copy logic "one level up" => into the buildThreads program.

I read somewhere, that chunk names inside magic comments get handled differently in v5.

e.g.:
```
        Neo.config.mainThreadAddons.forEach(addon => {
            if (addon !== 'AnalyticsByGoogle') {
                imports.push(import(/* webpackChunkName: 'src/main/addon/[request]' */ `./main/addon/${addon}.mjs`));
            }
        });
```

In v4 this only affected the (dist) dev env, in v5 i think it was for prod as well.
Will double-check if the chunk paths are still reasonable.

Did you check if the dist versions of some examples or apps still work with the changes?

There is one override for the main thread in place:

https://github.com/neomjs/neo/blob/dev/src/Main.mjs#L195
```
        if (window.webpackJsonp) {
            __webpack_require__.p = Neo.config.basePath.substring(6);
        }
```

We need to verify that this one still works (not nice, but mandatory, since the main thread can get picked up from different levels inside the folder hierarchy).

Will ping you tomorrow!

Best regards,
Tobias

**Comment by h1b9b** [comment]
Thanks for the tips, I went too fast for the upgrade yesterday. Sorry 🙏 

> Did you check if the WebpackShellPlugin still works? (buildThreads => App)
> The last time I tested it, it broke (quite a while ago though).
>
> It would not be a big deal in case the plugin does not support webpack v5. We don't really need the onBuildExit hook => we could move the docs resources copy logic "one level up" => into the buildThreads program.

I went for a simple replacement solution. So I replaced [WebpackShellPlugin](https://github.com/1337programming/webpack-shell-plugin) with [WebpackHookPlugin](https://github.com/tienne/webpack-hook-plugin). Cause `WebpackShellPlugin` didn't get an update since 2017 and `WebpackHookPlugin` uses the new webpack hook API

I also upgraded css/post-css loaders and their usage to have full webpack 5 support

There is warning on `html-webpack-plugin` that is fixed in the alpha version so I pushed a commit updating it but it can be removed until the next release is available cause the build is still ok with the warning. It should be release soon (cf [webpack 5 support issue](https://github.com/jantimon/html-webpack-plugin/issues/1527))


There a problem with the chunk names. The apps are still working.
But the chunks aren't split as before in the dev build and the indexes on the production build are different.
So I put the PR in WIP until I figure out what's going on and fix it 😅

**Comment by tobiu** [comment]
got time now to dive into your changes. will merge them first to get you into the contributors list.

File: buildScripts/docs/jsdocx.js (hunk 1 above)
Comment (tobiu) [reply buildScripts/docs/jsdocx.js:381]: Will create a new ticket for this one.

**Pull request merged**